trajopt_optimizers/trajopt_sqp/include/trajopt_sqp/qp_problem.h

```diff
@@ -0,0 +1,131 @@
+
+  /** @brief Sets up the problem and initializes matrices
+   * @param nlp
+   */
+  void init(ifopt::Problem& nlp);
+  /** @brief Run the full convexification routine. If in doubt, init(nlp) then convexify() */
+  void convexify();
+
+  /** @brief Called by convexify() - helper that updates the cost hessian (Currently does nothing since IFOPT doesn't
+   * support Hessians*/
+  void updateHessian();
+  /** @brief Called by convexify() - helper that updates the cost gradient including slack variables */
+  void updateGradient();
+  /** @brief Called by convexify() - helper that linearizes the constraints about the current point, storing the
+   * jacobian as the constraint matrix and adding slack variables.*/
+  void linearizeConstraints();
+  /** @brief Called by convexify() - helper that updates the NLP constraint bounds (top section) */
+  void updateNLPConstraintBounds();
+  /** @brief Called by convexify() - helper that updates the NLP variable bounds (middle section) */
+  void updateNLPVariableBounds();
+  /** @brief Called by convexify() - helper that updates the slack variable bounds (bottom section) */
+  void updateSlackVariableBounds();
+
+  /** @brief Evaluates the cost of the convexified function (ie using the stored gradient and hessian) at var_vals
+   * @param var_vals Point at which the convex cost is calculated. Should be size num_qp_vars
+   */
+  double evaluateTotalConvexCost(const Eigen::Ref<Eigen::VectorXd>& var_vals);
+  /** @brief TODO: This is unimplemented, but it will return the cost associated with each of the costs. Note this will
+   * be relatively computationally expensive, as we will have to loop through all the cost components in the problem and
+   * calculate their values manually.
+   * @param var_vals Point at which the convex cost is calculated. Should be size num_qp_vars
+   * @return Cost associated with each cost term in the problem (for debugging)
+   */
+  Eigen::VectorXd evaluateConvexCosts(const Eigen::Ref<Eigen::VectorXd>& var_vals);
+
+  /**
+   * @brief get the current NLP constraint violations. Values > 0 are violations
+   * @return Vector of constraint violations. Values > 0 are violations
+   */
+  Eigen::VectorXd getExactConstraintViolations();
+  /**
+   * @brief Uniformly scales the box size  (box_size_ = box_size_ * scale)
+   * @param scale Value by which the box size is scaled
+   */
+  void scaleBoxSize(double& scale);
+  /**
+   * @brief Sets the box size to the input vector
+   * @param box_size New box size
+   */
+  void setBoxSize(const Eigen::Ref<const Eigen::VectorXd>& box_size);
+  /**
+   * @brief Returns the box size
+   * @return The box size for each variable
+   */
+  Eigen::VectorXd getBoxSize() const;
+
+  /** @brief Prints all members to the terminal in a human readable form */
+  void print() const;
+
+protected:
+  ifopt::Problem* nlp_;
+
+  // TODO: I don't really want people to have write access to these. make const getters
+public:
+  Eigen::Index num_nlp_vars_;
```

Comment:
What class needs access to these? Maybe you can add it as a friend class so these do not need to be public.

Reply:
Made const getters for these.